---
title: 'CA2000: スコープを失う前にオブジェクトを破棄する (コード分析)'
description: 'コード分析ルール CA2000 の詳細: スコープを失う前にオブジェクトを破棄する'
ms.date: 05/14/2019
ms.topic: reference
f1_keywords:
- CA2000
- Dispose objects before losing scope
- DisposeObjectsBeforeLosingScope
helpviewer_keywords:
- CA2000
- DisposeObjectsBeforeLosingScope
author: gewarren
ms.author: gewarren
dev_langs:
- CSharp
- VB
ms.openlocfilehash: 2e8f2d58ae8f8799fde51cc15522277cbc90988a
ms.sourcegitcommit: ddf7edb67715a5b9a45e3dd44536dabc153c1de0
ms.translationtype: MT
ms.contentlocale: ja-JP
ms.lasthandoff: 02/06/2021
ms.locfileid: "99792428"
---
# <a name="ca2000-dispose-objects-before-losing-scope"></a>CA2000:スコープを失う前にオブジェクトを破棄

| | 値 |
|-|-|
| **ルール ID** |CA2000|
| **カテゴリ** |[信頼性](reliability-warnings.md)|
| **修正が中断または非対応になっています** |なし|

## <a name="cause"></a>原因

型のローカルオブジェクト <xref:System.IDisposable> が作成されますが、オブジェクトへのすべての参照がスコープ外になる前に、オブジェクトが破棄されることはありません。

既定では、このルールによってコードベース全体が分析されますが、これは [構成可能](#configure-code-to-analyze)です。

## <a name="rule-description"></a>規則の説明

破棄できるオブジェクトに対するすべての参照がスコープ外になる前に、オブジェクトが明示的に破棄されない場合、ガベージ コレクターでオブジェクトのファイナライザーが実行されるときに破棄されますが、タイミングは一定ではありません。 例外的なイベントが発生するとオブジェクトのファイナライザーを実行できないため、オブジェクトは明示的に破棄する必要があります。

## <a name="special-cases"></a>特殊なケース

オブジェクトが破棄されていない場合でも、次の型のローカルオブジェクトに対して Rule CA2000 は起動されません。

- <xref:System.IO.Stream?displayProperty=nameWithType>
- <xref:System.IO.StringReader?displayProperty=nameWithType>
- <xref:System.IO.TextReader?displayProperty=nameWithType>
- <xref:System.IO.TextWriter?displayProperty=nameWithType>
- <xref:System.Resources.IResourceReader?displayProperty=nameWithType>

これらの型のいずれかのオブジェクトをコンストラクターに渡し、それをフィールドに割り当てることは、新しく構築された型への *dispose の所有権の譲渡* を意味します。 つまり、新しく構築された型が、オブジェクトの破棄を担当するようになりました。 コードがこれらの型のいずれかのオブジェクトをコンストラクターに渡すと、オブジェクトへのすべての参照がスコープ外になる前に、オブジェクトが破棄されていない場合でも、規則 CA2000 の違反は発生しません。

## <a name="how-to-fix-violations"></a>違反の修正方法

この規則違反を修正するには、オブジェクトに対するすべての参照がスコープ外になる前に、そのオブジェクトで <xref:System.IDisposable.Dispose%2A> を呼び出します。

を実装するオブジェクトをラップするには、(Visual Basic の) [ `using` ステートメント](../../../csharp/language-reference/keywords/using-statement.md)を使用し [`Using`](../../../visual-basic/language-reference/statements/using-statement.md) <xref:System.IDisposable> ます。 この方法でラップされたオブジェクトは、ブロックの最後に自動的に破棄され `using` ます。 ただし、次の状況は、ステートメントを使用して処理したり、処理したりすることはできません `using` 。

- 破棄可能なオブジェクトを返すには、オブジェクトが `try/finally` ブロックの外側のブロック内に構築されている必要があり `using` ます。

- ステートメントのコンストラクターで、破棄可能なオブジェクトのメンバーを初期化しないでください `using` 。

- 1つの例外ハンドラーによって保護されているコンストラクターが [ `using` ステートメントの取得部分](../../../csharp/language-reference/keywords/using-statement.md)で入れ子になっている場合、外側のコンストラクターでエラーが発生すると、入れ子になったコンストラクターによって作成されたオブジェクトが閉じられることはありません。 次の例では、コンストラクターでエラーが発生すると、 <xref:System.IO.StreamReader> <xref:System.IO.FileStream> オブジェクトが閉じられない可能性があります。 CA2000 は、この場合、ルールの違反をフラグにします。

   ```csharp
   using (StreamReader sr = new StreamReader(new FileStream("C:\myfile.txt", FileMode.Create)))
   { ... }
   ```

- 動的オブジェクトでは、オブジェクトの dispose パターンを実装するために shadow オブジェクトを使用する必要があり <xref:System.IDisposable> ます。

## <a name="when-to-suppress-warnings"></a>警告を非表示にする場合

次の場合を除き、この規則からの警告を抑制しないでください。

- を呼び出すオブジェクトでメソッドを呼び出しました。 `Dispose` たとえば、 <xref:System.IO.Stream.Close%2A>
- 警告を発生させたメソッドは、 <xref:System.IDisposable> オブジェクトをラップするオブジェクトを返します。
- 割り当てメソッドに dispose の所有権がありません。つまり、オブジェクトを破棄する責任は、メソッドで作成され、呼び出し元に返される別のオブジェクトまたはラッパーに転送されます。

## <a name="configure-code-to-analyze"></a>分析するコードの構成

次のオプションを使用して、この規則を実行するコードベースの部分を構成します。

- [特定のシンボルを除外する](#exclude-specific-symbols)
- [特定の型とその派生型を除外する](#exclude-specific-types-and-their-derived-types)

これらのオプションは、この規則、すべての規則、またはこのカテゴリのすべての規則 ([信頼性](reliability-warnings.md)) に対して構成できます。 詳細については、「 [コード品質ルールの構成オプション](../code-quality-rule-options.md)」を参照してください。

[!INCLUDE[excluded-symbol-names](~/includes/code-analysis/excluded-symbol-names.md)]

[!INCLUDE[excluded-type-names-with-derived-types](~/includes/code-analysis/excluded-type-names-with-derived-types.md)]

## <a name="related-rules"></a>関連するルール

- [CA2213:破棄可能なフィールドは破棄されなければなりません](ca2213.md)

## <a name="example-1"></a>例 1

破棄可能なオブジェクトを返すメソッドを実装する場合は、catch ブロックのない try/finally ブロックを使用して、オブジェクトが破棄されていることを確認してください。 try/finally ブロックを使用することによって、障害点での例外の発生が可能になり、そのオブジェクトが確実に破棄されます。

OpenPort1 メソッドでは、ISerializable オブジェクトの SerialPort を開くための呼び出し、または SomeMethod の呼び出しが失敗する可能性があります。 この実装では CA2000 警告は発生しません。

OpenPort2 メソッドでは、次の 2 つの SerialPort オブジェクトが宣言され、null に設定されます。

- `tempPort`。メソッド操作が成功しているかどうかをテストするのに使用されます。

- `port`。メソッドの戻り値に使用されます。

`tempPort` は、`try` ブロックで構築され、開かれます。その他必要な作業も同じ `try` ブロック内で実行されます。 `try` ブロックの最後に、開かれたポートが `port` オブジェクトに割り当てられ、このオブジェクトが返されます。`tempPort` オブジェクトは `null` に設定されます。

`finally` ブロックは `tempPort` 値をチェックします。 null でない場合、メソッド内の操作は失敗しています。`tempPort` は閉じられ、すべてのリソースが解放されます。 返されるオブジェクトには、メソッド操作が成功した場合、開かれた SerialPort オブジェクトが含まれます。メソッドの操作が失敗した場合は null になります。

```csharp
public SerialPort OpenPort1(string portName)
{
   SerialPort port = new SerialPort(portName);
   port.Open();  //CA2000 fires because this might throw
   SomeMethod(); //Other method operations can fail
   return port;
}

public SerialPort OpenPort2(string portName)
{
   SerialPort tempPort = null;
   SerialPort port = null;
   try
   {
      tempPort = new SerialPort(portName);
      tempPort.Open();
      SomeMethod();
      //Add any other methods above this line
      port = tempPort;
      tempPort = null;

   }
   finally
   {
      if (tempPort != null)
      {
         tempPort.Close();
      }
   }
   return port;
}
```

```vb
Public Function OpenPort1(ByVal PortName As String) As SerialPort

   Dim port As New SerialPort(PortName)
   port.Open()    'CA2000 fires because this might throw
   SomeMethod()   'Other method operations can fail
   Return port

End Function

Public Function OpenPort2(ByVal PortName As String) As SerialPort

   Dim tempPort As SerialPort = Nothing
   Dim port As SerialPort = Nothing

   Try
      tempPort = New SerialPort(PortName)
      tempPort.Open()
      SomeMethod()
      'Add any other methods above this line
      port = tempPort
      tempPort = Nothing

   Finally
      If Not tempPort Is Nothing Then
         tempPort.Close()
      End If

   End Try

   Return port

End Function
```

## <a name="example-2"></a>例 2

既定では、Visual Basic コンパイラは、すべての算術演算子のオーバーフローをチェックします。 そのため、いずれかの Visual Basic 算術演算子で <xref:System.OverflowException> がスローされる可能性があります。 これにより、CA2000 のような予期しない規則違反が発生する場合があります。 たとえば、次の CreateReader1 関数では、Visual Basic コンパイラが加算に対するオーバーフロー チェックを実行し、それが例外をスローすると StreamReader が破棄されなくなるので、CA2000 違反が発生します。

これを修正するには、プロジェクトで Visual Basic コンパイラによるオーバーフロー チェックの実施を無効にするか、または次の CreateReader2 関数のようにコードを変更します。

オーバーフローチェックの出力を無効にするには、ソリューションエクスプローラーでプロジェクト名を右クリックし、[ **プロパティ**] をクリックします。 [ **コンパイル**] をクリックし、[ **詳細コンパイルオプション**] をクリックして、[ **整数オーバーフローのチェックを削除**] をオンにします。

:::code language="vb" source="snippets/vb/all-rules/ca2000-dispose-objects-before-losing-scope-vboverflow_1.vb":::

## <a name="see-also"></a>関連項目

- <xref:System.IDisposable>
- [Dispose パターン](../../../standard/garbage-collection/implementing-dispose.md)
